### ISSUE TITLE
Newly-compiled macros not available in the context already in use

### ISSUE THREAD

**Issue by jobarr-amzn** [amazon-ion/ion-rust#817]
Macros work and it's neat!
```plain
❯ cat hello.ion
$ion_1_1
$ion_encoding::(
  (macro_table
    (macro hello ($name) (make_string "hello " $name))
  )
)

(:hello "world")
```
```plain
❯ ion cat hello.ion
"hello world"
```

But if I reference in an encoding context a macro defined earlier in that same context it does not work:
```plain
❯ cat broken.ion
$ion_1_1
$ion_encoding::(
  (macro_table
    (macro hello ($name) (make_string "hello " $name))
    (macro hello_world (:hello "world"))
  )
)

(:hello_world)
```
```plain
❯ ion cat broken.ion
Error: could not find macro with id LocalName("hello")
        offset=119
        buffer head=<(:hello "world"))
  )
)
```

The newly compiled macro gets added to the macro_table being built up, but that means it isn't available in the context already in use:
https://github.com/amazon-ion/ion-rust/blob/0e941e3175c2cf6b5a40a0b38692b744f86b2f10/src/lazy/system_reader.rs#L364-L372

**Comment by jobarr-amzn** [comment]
This doesn't seem right to me, that `match_e_expression` should be trying to look up macro definitions in order to parse the one in front of it: https://github.com/amazon-ion/ion-rust/blob/0e941e3175c2cf6b5a40a0b38692b744f86b2f10/src/lazy/text/buffer.rs#L1103-L1113

I would think that (allowing for context) `(:foo bar baz qux)` would be understood as an attempt to invoke a macro named `foo` with arguments `foo`, `baz`, and `qux`- no matter what the context was. Right? The thing may be semantically meaningless but the syntax is clear.

**Comment by jobarr-amzn** [comment]
> This doesn't seem right to me, that `match_e_expression` should be trying to look up macro definitions in order to parse the one in front of it:
> 
> https://github.com/amazon-ion/ion-rust/blob/0e941e3175c2cf6b5a40a0b38692b744f86b2f10/src/lazy/text/buffer.rs#L1103-L1113
> 
> I would think that (allowing for context) `(:foo bar baz qux)` would be understood as an attempt to invoke a macro named `foo` with arguments `foo`, `baz`, and `qux`- no matter what the context was. Right? The thing may be semantically meaningless but the syntax is clear.

This is misleading- the problem is not here. I was chasing this direction because my *input* was wrong. Test input should be:
```
$ion_1_1
$ion_encoding::(
  (macro_table
    (macro hello ($name) (make_string "hello " $name))
    (macro hello_world () (hello "world"))
  )
)

(:hello_world)
```

**Comment by jobarr-amzn** [comment]
Fixed by #822
